### PR TITLE
[FIX] payment: Better error message for Stripe session

### PR DESCRIPTION
**Pull request by lse-odoo**
Before this commit:
When opening a Stripe session, if an error happened, a basic error message was displayed with no relevant information in it.
You can reproduce it by adding some space character in the partner email for instances.

With this commit:
Some details on the errors reason is given into the payment form and more detailed into the sale order/invoice chatter.

OPW-2225881
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by rim-odoo** [comment]
@AntoineVDV if you have some time, can you have a look at this? as @bouvyd shamelessly left us ;)

@lse-odoo I would target 13.5, so that we don't break existing installations

File: addons/payment_stripe/i18n/payment_stripe.pot [hunk 1]
```diff
@@ -167,6 +167,30 @@ msgstr ""
...
+#. module: payment_stripe
+#: code:addons/payment_stripe/models/payment.py
+#, python-format
+msgid "Error while creating Stripe session: %s."
+msgstr ""
+
+#. module: payment_stripe
+#: code:addons/payment_stripe/models/payment.py
+#, python-format
+msgid "Please double check that the user email is formatted correctly : '%s'"
```
Comment (rim-odoo) [inline addons/payment_stripe/i18n/payment_stripe.pot:179]: Grammar nitpicking
```suggestion
msgid "Please double-check that the user email is formatted correctly : '%s'"
```

File: addons/payment_stripe/i18n/payment_stripe.pot [hunk 2]
```diff
@@ -167,6 +167,30 @@ msgstr ""
...
+#. module: payment_stripe
+#: code:addons/payment_stripe/models/payment.py
+#, python-format
+msgid "Stripe error code: '%s'"
+msgstr ""
+
+#. module: payment_stripe
+#: code:addons/payment_stripe/models/payment.py
+#, python-format
+msgid "See documentation at: %s"
```
Comment (rim-odoo) [inline addons/payment_stripe/i18n/payment_stripe.pot:191]:
```suggestion
msgid "See Stripe documentation page at: %s"
```

File: addons/payment_stripe_sca/models/payment.py [hunk 3]
```diff
@@ -52,6 +52,8 @@ def _stripe_request(self, url, data=False, method="POST"):
             "AUTHORIZATION": "Bearer %s" % self.sudo().stripe_secret_key,
             "Stripe-Version": "2019-05-16",  # SetupIntent need a specific version
         }
+        if data and data.get("customer_email"):
+            data["customer_email"] = data["customer_email"].strip()
```
Comment (rim-odoo) [inline addons/payment_stripe_sca/models/payment.py:56]: If this line addresses a bug with un-stripped email address, it should be fixed in a separate PR IMHO

File: addons/payment_stripe_sca/models/payment.py [hunk 4]
```diff
@@ -70,15 +72,42 @@ def _stripe_request(self, url, data=False, method="POST"):
...
-                .search([("reference", "=", kwargs["client_reference_id"])])
-            )
-            tx.stripe_payment_intent = resp["payment_intent"]
-        return resp["id"]
+        resp = self._stripe_request('checkout/sessions', kwargs)
+        tx = False
+        if kwargs.get('client_reference_id'):
+            tx = self.env['payment.transaction'].sudo().search([('reference', '=', kwargs['client_reference_id'])])
+            if resp.get('payment_intent'):
+                tx.stripe_payment_intent = resp['payment_intent']
```
Comment (rim-odoo) [inline addons/payment_stripe_sca/models/payment.py:80]: How can I test this flow? I guess you are trying to fix a case where there is no `payment_intent` :thinking:

Comment (rim-odoo) [reply addons/payment_stripe_sca/models/payment.py:80]: Same remark: if you are trying to fix a bug => separate PR than this one that ought to log better error messages

File: addons/payment_stripe_sca/models/payment.py [hunk 5]
```diff
@@ -70,15 +70,42 @@ def _stripe_request(self, url, data=False, method="POST"):
...
+        error = resp.get("error")
+        if error:
+            _logger.error("Stripe session creation error %s" % error)
+
+            error_msg = (_("Error while creating Stripe session: %s.") % (error.get("message", "No message")))
+
+            error_code = error.get("code")
+            customer_email = kwargs.get("customer_email")
+            if error_code == "email_invalid" and customer_email:
+                error_msg += "<br>" + (_("Please double check that the user email is formatted correctly : '%s'") % customer_email)
```
Comment (rim-odoo) [inline addons/payment_stripe_sca/models/payment.py:91]:
```suggestion
                error_msg += "<br>" + (_("Please double-check that the user email is formatted correctly : '%s'") % customer_email)
```

File: addons/payment_stripe_sca/models/payment.py [hunk 6]
```diff
@@ -70,15 +70,42 @@ def _stripe_request(self, url, data=False, method="POST"):
...
+
+            if tx:
+                targets = {tx.invoice_ids, tx.sale_order_ids}
+
+                error_message_details = ""
+                if error_code:
+                    error_message_details += "<br>" + _("Stripe error code: '%s'") % error_code
+                error_message_details += "<br>" + _("See documentation at: %s") % error.get("doc_url", "https://stripe.com/docs/error-codes")
```
Comment (rim-odoo) [inline addons/payment_stripe_sca/models/payment.py:99]:
```suggestion
                error_message_details += "<br>" + _("See Stripe documentation page at: %s") % error.get("doc_url", "https://stripe.com/docs/error-codes")
```

File: addons/payment_stripe_sca/models/payment.py [hunk 7]
```diff
@@ -70,15 +70,42 @@ def _stripe_request(self, url, data=False, method="POST"):
...
+
+                self.env.cr.rollback()
+                for target in targets:
+                    for order in target:
+                        order.message_post(
+                            body=error_msg + error_message_details,
+                        )
+                self.env.cr.commit()
```
Comment (rim-odoo) [inline addons/payment_stripe_sca/models/payment.py:107]: Instead of "roll-backing" then committing, you can also work with a separate cursor, something like this:
https://github.com/odoo/odoo/blob/13.0/addons/delivery/models/delivery_carrier.py#L210-L225

(by the way, the `ValidationError` is going to force a roll-back)

**Review by rim-odoo — CHANGES_REQUESTED** [review]

**Comment by AntoineVDV** [comment]
@lse-odoo Can I already have a look at it, or should I wait for additional changes?

**Comment by lse-odoo** [comment]
By trying to reproduce the same way that I did originally the issue does not appear anymore. I think https://github.com/odoo/odoo/pull/50908 did fix it in another (more clever) way.
We can close this PR IMO.
